File: c#ListUserVoicemailParms/listUserVoicemailParameters.cs
using System;
using System.Globalization;
using System.IO;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Threading.Tasks;
using CsvHelper;
using CsvHelper.Configuration;

class Program
{
    // ----- Configuration -----
    // Prefer ENV VARS; falls back to hardcoded defaults if you set them here.
    // Example: setx WEBEX_BASE https://webexapis.com
    //          setx WEBEX_TOKEN <YOUR_ADMIN_TOKEN>
    //          setx WEBEX_ORG_ID <YOUR_ORG_ID>
    private static readonly string BASE =
        Environment.GetEnvironmentVariable("WEBEX_BASE") ?? "https://webexapis.com";

    // If you insist on hard-coding, put the token/orgId here — but ENV VARS are safer.
    
    private static readonly string TOKEN =
        Environment.GetEnvironmentVariable("WEBEX_TOKEN") ?? ""; // "<YOUR TOKEN>";

    private static readonly string ORG_ID =
        Environment.GetEnvironmentVariable("WEBEX_ORG_ID") ?? ""; // "<YOUR ORG_ID>";
    

    private static readonly HttpClient http = new HttpClient
    {
        Timeout = TimeSpan.FromSeconds(30)
    };

    static async Task<int> Main(string[] args)
    {
        // ----- Inputs/Outputs -----
        string inputCsv = "email2personID.csv";
        string outputCsv = "userVoicemailParameters.csv";

        if (string.IsNullOrWhiteSpace(TOKEN))
        {
            Console.Error.WriteLine("ERROR: Set the WEBEX_TOKEN environment variable to a valid admin token.");
            return 1;
        }

        if (!File.Exists(inputCsv))
        {
            Console.Error.WriteLine($"CSV file not found: {inputCsv}");
            return 1;
        }

        // ----- HTTP headers -----
        http.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", TOKEN);
        http.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

        // ----- CSV setup -----
        var csvConfig = new CsvConfiguration(CultureInfo.InvariantCulture)
        {
            HasHeaderRecord = true,
            TrimOptions = TrimOptions.Trim,
            BadDataFound = null
        };

        // Prepare writer with headers that mirror your Python fields
        using var writer = new StreamWriter(outputCsv);
        using var csvOut = new CsvWriter(writer, csvConfig);
        csvOut.WriteField("email");
        csvOut.WriteField("personid");
        csvOut.WriteField("enabled");
        csvOut.WriteField("storageType");
        csvOut.WriteField("mwiEnabled");
        csvOut.WriteField("externalEmail");
        csvOut.WriteField("emailCopyEnabled");
        csvOut.WriteField("emailCopyId");
        await csvOut.NextRecordAsync();

        // Read input and process rows
        using var reader = new StreamReader(inputCsv);
        using var csvIn = new CsvReader(reader, csvConfig);
        await foreach (var row in csvIn.GetRecordsAsync<dynamic>())
        {
            // dynamic row behaves like IDictionary<string, object>
            var dict = (System.Collections.Generic.IDictionary<string, object>)row;

            string email = (dict.TryGetValue("email", out var e) ? Convert.ToString(e) : "")?.Trim() ?? "";
            string personId = (dict.TryGetValue("personid", out var p) ? Convert.ToString(p) : "")?.Trim() ?? "";

            if (string.IsNullOrWhiteSpace(personId))
            {
                Console.WriteLine($"⚠️  Skipping missing personid for {email}");
                continue;
            }

            var data = await GetVoicemailForPersonAsync(personId);
            if (data is null) continue;

            // Extract the same fields as the Python script
            bool? enabled = data["enabled"]?.GetValue<bool?>();

            var messageStorage = data["messageStorage"] as JsonObject;
            string? storageType = messageStorage?["storageType"]?.GetValue<string?>();
            bool? mwiEnabled = messageStorage?["mwiEnabled"]?.GetValue<bool?>();
            string? externalEmail = messageStorage?["externalEmail"]?.GetValue<string?>();

            var emailCopy = data["emailCopyOfMessage"] as JsonObject;
            bool? emailCopyEnabled = emailCopy?["enabled"]?.GetValue<bool?>();
            string? emailCopyId = emailCopy?["emailId"]?.GetValue<string?>();

            // Write CSV row
            csvOut.WriteField(email);
            csvOut.WriteField(personId);
            csvOut.WriteField(NullableToString(enabled));
            csvOut.WriteField(storageType ?? "");
            csvOut.WriteField(NullableToString(mwiEnabled));
            csvOut.WriteField(externalEmail ?? "");
            csvOut.WriteField(NullableToString(emailCopyEnabled));
            csvOut.WriteField(emailCopyId ?? "");
            await csvOut.NextRecordAsync();

            // Console output
            Console.WriteLine($"\n📧 {email} ({personId})");
            Console.WriteLine($"  Voicemail Enabled: {NullableToString(enabled)}");
            Console.WriteLine($"  Storage Type: {storageType}");
            Console.WriteLine($"  MWI Enabled: {NullableToString(mwiEnabled)}");
            Console.WriteLine($"  External Email: {externalEmail}");
            Console.WriteLine($"  Email Copy Enabled: {NullableToString(emailCopyEnabled)}");
            Console.WriteLine($"  Email Copy ID: {emailCopyId}");
        }

        Console.WriteLine($"\n✅ All results written to {outputCsv}");
        return 0;
    }

    private static async Task<JsonObject?> GetVoicemailForPersonAsync(string personId)
    {
        var url = $"{BASE}/v1/people/{Uri.EscapeDataString(personId)}/features/voicemail";
        if (!string.IsNullOrWhiteSpace(ORG_ID))
        {
            url += $"?orgId={Uri.EscapeDataString(ORG_ID)}";
        }

        try
        {
            using var resp = await http.GetAsync(url);
            var payload = await resp.Content.ReadAsStringAsync();

            if (resp.IsSuccessStatusCode)
            {
                var node = JsonNode.Parse(payload) as JsonObject;
                return node;
            }
            else
            {
                Console.WriteLine($"❌ {personId}: {(int)resp.StatusCode} {payload}");
                return null;
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"❌ {personId}: Exception {ex.Message}");
            return null;
        }
    }

    private static string NullableToString(bool? b) => b.HasValue ? b.Value.ToString() : "";
}

File: vmParamsListPy/listUserVoicemailParameters.py
import os
import csv
import sys
import requests

# ----- Environment setup -----
BASE = os.getenv("WEBEX_BASE", "https://webexapis.com")
TOKEN = os.getenv("WEBEX_TOKEN")  # Set this before running
ORG_ID = os.getenv("WEBEX_ORG_ID")
# TOKEN = "MTkxZDZjODctODA5My00ZWZmLThlZmUtMzNjZDkzM2I5NjkxMGVmYTg0MjAtYzRm_PF84_ebc31646-bc26-4a20-8500-c3030ebd6a52"
# ORG_ID = "ebc31646-bc26-4a20-8500-c3030ebd6a52"

if not TOKEN:
    sys.exit("ERROR: Set the WEBEX_TOKEN environment variable to a valid admin token.")

# ----- HTTP headers -----
def hdrs():
    return {
        "Authorization": f"Bearer {TOKEN}",
        "Content-Type": "application/json"
    }

# ----- Get voicemail parameters -----
def get_voicemail_for_person(person_id: str):
    params = {"orgId": ORG_ID} if ORG_ID else {}
    url = f"{BASE}/v1/people/{person_id}/features/voicemail"
    r = requests.get(url, headers=hdrs(), params=params, timeout=30)
    if r.status_code == 200:
        return r.json()
    else:
        print(f"❌ {person_id}: {r.status_code} {r.text}")
        return None

# ----- Main workflow -----
def main():
    input_csv = "../email2personID.csv"
    output_csv = "userVoicemailParameters.csv"

    if not os.path.exists(input_csv):
        sys.exit(f"CSV file not found: {input_csv}")

    with open(input_csv, newline="") as infile, open(output_csv, "w", newline="") as outfile:
        reader = csv.DictReader(infile)
        fieldnames = ["email", "personid", "enabled", "storageType", "mwiEnabled",
                      "externalEmail", "emailCopyEnabled", "emailCopyId"]
        writer = csv.DictWriter(outfile, fieldnames=fieldnames)
        writer.writeheader()

        for row in reader:
            email = row.get("email", "").strip()
            person_id = row.get("personid", "").strip()
            if not person_id:
                print(f"⚠️  Skipping missing personid for {email}")
                continue

            data = get_voicemail_for_person(person_id)
            if not data:
                continue

            # Extract key voicemail parameters
            enabled = data.get("enabled")
            messageStorage = data.get("messageStorage", {})
            emailCopy = data.get("emailCopyOfMessage", {})

            record = {
                "email": email,
                "personid": person_id,
                "enabled": enabled,
                "storageType": messageStorage.get("storageType"),
                "mwiEnabled": messageStorage.get("mwiEnabled"),
                "externalEmail": messageStorage.get("externalEmail"),
                "emailCopyEnabled": emailCopy.get("enabled"),
                "emailCopyId": emailCopy.get("emailId")
            }

            writer.writerow(record)

            # Console output
            print(f"\n📧 {email} ({person_id})")
            print(f"  Voicemail Enabled: {enabled}")
            print(f"  Storage Type: {record['storageType']}")
            print(f"  MWI Enabled: {record['mwiEnabled']}")
            print(f"  External Email: {record['externalEmail']}")
            print(f"  Email Copy Enabled: {record['emailCopyEnabled']}")
            print(f"  Email Copy ID: {record['emailCopyId']}")

    print(f"\n✅ All results written to {output_csv}")

# ----- Entrypoint -----
if __name__ == "__main__":
    main()
